File: sdks/csharp/PgChangeFeed.Client/Nats/Models/Change.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PgChangeFeed.Client.Nats.Models;

/// <summary>
/// One change delivered on the NATS stream — the ten fields the server writes
/// into the JSON payload of a message on
/// <c>cdc.stream.&lt;source_id&gt;.&lt;schema&gt;.&lt;table&gt;</c>, one JSON
/// object per change. <see cref="OldImage"/>/<see cref="NewImage"/> carry the
/// row image as an embedded JSON value or <c>null</c> when absent; kept as
/// <see cref="JsonElement"/> rather than a fixed shape, because the row
/// image's own shape depends on the captured table (same as
/// <see cref="PgChangeFeed.Client.Sse.Models.Change"/>).
///
/// The ten fields are the same as those of the SSE event, but the type is its
/// own: the HTTP read returns <see cref="PgChangeFeed.Client.Http.Models.Change"/>
/// (thirteen fields including <c>commit_position</c>/<c>committed_at</c>/<c>origin</c>),
/// the gRPC client returns the generated <c>Change</c> message, the SSE client
/// returns <see cref="PgChangeFeed.Client.Sse.Models.Change"/> — four clients,
/// four independent types that share most field names.
/// </summary>
public sealed record Change(
    [property: JsonPropertyName("change_id")] string ChangeId,
    [property: JsonPropertyName("transaction_id")] string TransactionId,
    [property: JsonPropertyName("source_table_id")] string SourceTableId,
    [property: JsonPropertyName("sequence")] long Sequence,
    [property: JsonPropertyName("operation")] string Operation,
    [property: JsonPropertyName("old_image")] JsonElement? OldImage,
    [property: JsonPropertyName("new_image")] JsonElement? NewImage,
    [property: JsonPropertyName("schema_version")] string SchemaVersion,
    [property: JsonPropertyName("schema")] string Schema,
    [property: JsonPropertyName("table")] string Table);

File: sdks/kotlin/pgchangefeed-kotlin/src/main/kotlin/io/github/pt9912/pgchangefeed/nats/model/Change.kt
package io.github.pt9912.pgchangefeed.nats.model

import com.google.gson.JsonElement
import com.google.gson.annotations.SerializedName

/**
 * One change delivered on the NATS stream — the ten fields the server writes
 * into a message's JSON payload, the same as those of the SSE message
 * ([io.github.pt9912.pgchangefeed.sse.model.Change]), carried over NATS instead
 * of HTTP. [oldImage]/[newImage] carry the row image as an embedded JSON value
 * or a JSON `null` when absent (`INSERT` has no `old_image`, `DELETE` has no
 * `new_image`); kept as [JsonElement] rather than a fixed shape for the same
 * reason as [io.github.pt9912.pgchangefeed.http.model.Change]'s corresponding
 * fields — the row image's own shape depends on the captured table.
 *
 * It is its own type, not shared with the other clients: the SSE client
 * returns [io.github.pt9912.pgchangefeed.sse.model.Change], the HTTP read
 * returns [io.github.pt9912.pgchangefeed.http.model.Change], and the gRPC
 * client returns the generated `cdc.stream.v1.Changestream.Change` message —
 * separate types that share most field names.
 */
data class Change(
    @SerializedName("change_id") val changeId: String,
    @SerializedName("transaction_id") val transactionId: String,
    @SerializedName("source_table_id") val sourceTableId: String,
    @SerializedName("sequence") val sequence: Long,
    @SerializedName("operation") val operation: String,
    @SerializedName("old_image") val oldImage: JsonElement?,
    @SerializedName("new_image") val newImage: JsonElement?,
    @SerializedName("schema_version") val schemaVersion: String,
    @SerializedName("schema") val schema: String,
    @SerializedName("table") val table: String,
)
